### PR TITLE
add a transformer suite in TP/SP tests

### LINKED ISSUE #115309: [distributed] add a "small" transformer model to the test suite for testing usage
We can add a "small" Transformer model to our test suite to test out our existing FSDP/TP/SP APIs to ensure no future breakge and give easy to read example about how to apply our APIs directly in tests.

### PR DESCRIPTION
**Pull request by tianyu-l**
This is to address issue #115309.

Test plan
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
`python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115530

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

### CODE REVIEW

**Review by wanchaol — COMMENTED** [review]
Mostly looks good! have a few comments inlined, once those are addressed I think we are good to go!

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 1]
```diff
@@ -152,6 +284,10 @@ def test_mlp_inference(self):
         with torch.inference_mode():
             self._test_mlp_inference(device_mesh)
 
+    @with_comms
+    @parametrize("is_seq_parallel", [True, False])
+    def test_transformer_training(self, is_seq_parallel):
+        self._test_transformer_training_e2e(is_seq_parallel=is_seq_parallel)
```
[the remark this answers is not in the document]
Comment (tianyu-l) [reply test/distributed/tensor/parallel/test_tp_examples.py]: Removed. I was just following the style of the other two tests in this file.

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 2]
```diff
@@ -152,6 +158,147 @@ def test_mlp_inference(self):
...
+                # parallelize_plan[f"layers.{i}.attention"] = PrepareModuleInput(
+                #     input_layouts=Shard(1),
+                #     desired_input_layouts=Replicate(),
+                # )
+                parallelize_module(
+                    model_tp.layers[i].attention,
+                    device_mesh,
+                    PrepareModuleInput(input_layouts=Shard(1), desired_input_layouts=Replicate())
+                )
+            parallelize_module(model_tp.layers[i].attention.wq, device_mesh, ColwiseParallel())
```
Comment (wanchaol) [inline test/distributed/tensor/parallel/test_tp_examples.py:209]: hmmm. I think we should have a separate parallelize_plan for each transformer layer defined as a dict of fqn -> parallel styles. Calling parallelize_module so many times is not preferred I think. 

Each layer we can have a parallelize_plan then a single `parallelize_module` call, and then the first two embedding layer and the final projection layer can call `paralllelize_module`.

Comment (tianyu-l) [reply test/distributed/tensor/parallel/test_tp_examples.py:209]: Oh I see. I misunderstood your previous comment; in fact I also thought this way of calling is not good. I wonder in general what principles user should follow when parallelizing their model. E.g. should one always try to group submodules within a bigger module into a single parallelize_plan, as long as the plan does grow very big? What are the concerns of not having a super giant single plan, like what I did earlier?

Comment (wanchaol) [reply test/distributed/tensor/parallel/test_tp_examples.py]: I think a super giant plan also works, it's just that when we have multiple layers with the same plan, things become a bit redundant, so I was thinking for transformer that have repeated sharding plan, per-layer sharding plan makes things more clean

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for working on this!

**Pull request closed without merging**